File: frontend/src/model/message.ts
import {Admin} from "./user/admin";
import {Professor} from "./user/professor";
import {Student} from "./user/student";

export class Message {
  messageId: number | undefined;
  message: string;
  timestamp: string;

  professorId: number;
  studentId: number;
  adminId: number;
  conversationId: number;

  admin: Admin | undefined;
  professor: Professor | undefined;
  student: Student | undefined;

  constructor(message: string, timestamp: string, professorId: number, studentId: number, adminId: number, conversationId: number, messageId?: number) {
    this.messageId = messageId;
    this.message = message;
    this.timestamp = timestamp;
    this.professorId = professorId;
    this.studentId = studentId;
    this.adminId = adminId;
    this.conversationId = conversationId;
  }

  static fromJson(json: Message): Message {
    let message = new Message(json.message, json.timestamp,
      json.professorId, json.studentId, json.adminId,
      json.conversationId, json.messageId);

    if(json.admin != undefined) message.admin =  Admin.fromJson(json.admin);
    if(json.professor != undefined) message.professor = Professor.fromJson(json.professor);
    if(json.student != undefined) message.student = Student.fromJson(json.student);

    return message;
  }

  static initializeMessages(jsonMessages: Message[]): Message[] {
    let messages: Message[] = [];
    if(jsonMessages != undefined) {
      for (let jsonMessage of jsonMessages) {
        messages.push(Message.fromJson(jsonMessage));
      }
    }
    return messages;
  }
}

File: backend/src/main/java/arbremojo/backend/model/Message.java
package arbremojo.backend.model;

import arbremojo.backend.model.user.Admin;
import arbremojo.backend.model.user.Professor;
import arbremojo.backend.model.user.Student;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.ColumnDefault;

import java.security.Timestamp;
import java.util.Date;

@Entity
@Table(name = "message")
@Setter
@Getter
@ToString
@AllArgsConstructor
@NoArgsConstructor
public class Message {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "message_id", nullable = false)
    private Integer messageId;
    @Column(name = "message", nullable = false)
    private String message;
    @Column(name = "timestamp", nullable = false)
    @Temporal(TemporalType.TIMESTAMP)
    @ColumnDefault("CURRENT_TIMESTAMP")
    private Date timestamp;

    @Column(name = "professor_id", nullable = false)
    private Integer professorId;

    @Column(name = "student_id", nullable = false)
    private Integer studentId;

    @Column(name = "admin_id", nullable = false)
    private Integer adminId;

    @Column(name = "conversation_id", nullable = false)
    private Integer conversationId;

    @Transient
    private Professor professor;

    @Transient
    private Student student;

    @Transient
    private Admin admin;
}
